packages/util-waiter/src/waiter.ts

```diff
@@ -55,4 +61,28 @@ export enum WaiterState {
 
 export type WaiterResult = {
   state: WaiterState;
+
+  /**
+   * (optional) Indicates a reason for why a waiter has reached its state.
+   */
+  reason?: any;
+};
+
+/**
+ * Handles and throws exceptions resulting from the waiterResult
+ * @param result WaiterResult
+ */
+export const checkExceptions = (result: WaiterResult): WaiterResult => {
+  if (result.state === WaiterState.ABORTED) {
+    throw new Error(
+      `${JSON.stringify({
+        ...result,
+        name: "AbortError",
+        reason: "Request was aborted",
+      })}`
+    );
```

Comment:
This command just stringifies the structure. The error.name is still "Error" and not "AbortError" as encouraged in [AbortController spec](https://dom.spec.whatwg.org/#aborting-ongoing-activities)

> APIs that rely upon AbortController are encouraged to respond to abort() by rejecting any unsettled promise with a new "AbortError" DOMException.

Verified by running the following in JavaScript console:

```console
new Error(
  `${JSON.stringify({
    name: "AbortError",
    reason: "Request was aborted",
  })}`
).name;
"Error"
```

Reply:
Also, emit a TimeoutError in case of Timeout - as done in http-handler for example: https://github.com/aws/aws-sdk-js-v3/blob/2384e2428879d72f67fcf8c4cb06c71ec59189df/packages/fetch-http-handler/src/request-timeout.ts#L5-L6

Reply:
... To be fair - this is probably best solved by creating (or importing) new TimeoutError or AbortError objects that themselves extend Error. I dont feel great about jerry rigging abortError or timeout error here. Do you know if we have existing AbortError or TimeoutError objects? 

I don't think throwing a domexception is appropriate in all the contexts here either

Reply:
okay sync'd offline. Just adding here